text/0000-union-initialization-and-drop.md

```diff
@@ -0,0 +1,374 @@
+- Feature Name: union_initialization_and_drop
+- Start Date: 2018-08-03
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Unions do not allow fields of types that require drop glue (the code that is
+automatically run when a variables goes out of scope: recursively dropping the
+variable and all its fields), but they may still `impl Drop` themselves.  We
+specify when one may move out of a union field and when the union's `drop` is
+called.  To avoid undesired implicit calls of drop, we also restrict the use of
+`DerefMut` when unions are involved.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, it is unstable to have a non-`Copy` field in the union.  The main
+reason for this is that having fields which need drop glue raises some hard
+questions about whether to call that drop glue when assigning a union field, and
+how to make programming with such unions less of a time bomb (triggered by
+accidentally dropping data one meant to just overwrite).  Not much progress has
+been made on stabilizing the unstable union features.  This RFC proposes a route
+forwards that side-steps the time bomb: Do not allow fields with drop glue.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Union Definition
+
+When defining a union, it is a hard error to use a field type that requires drop glue.
+Examples:
+```rust
+// Accepted
+union Example1<T> {
+    // `ManuallyDrop<T>` never has drop glue, even if `T` does.
+    f1: ManuallyDrop<T>,
+    // `RefCell<i32>` is a fully known type, and does not have drop glue.
+    f2: RefCell<i32>,
+}
+union Example2<T: Copy> {
+    // `Copy` types never have drop glue.
+    f1: T,
+}
+trait Trait3 { type Assoc: Copy; }
+union Example3<T: Trait3> {
+    // `T::Assoc` is `Copy` and hence cannot have drop glue.
+    f1: T::Assoc,
+}
+
+// Rejected
+union Example4<T> {
+    // `T` might have drop glue, and then `Cell<T>` would as well.
```

Comment:
Should this mention `RefCell<T>`?

Reply:
Why that? I didn't intend to list every time with interior mutability.

Reply:
Specifically because the line below is `f1: RefCell<T>`. I assumed this comment was here to explain why, given that line, this struct is rejected.

Reply:
Oh, I made a typo and didn't use the same type on both sides. Thanks for pointing that out! (GitHub's diff viw for comments embedded in the discussions is *so* bad...)

Reply:
No problem - sorry I didn't point it out better initially! I'm still not entirely used to how github displays these things either.